refactor: replace config system with structopt

domodwyer opened the pull request:
Replaces the hand-rolled config system with a StructOpt managed config struct.

I've got most of it ported across, but the interaction between all the logging
config bits is complex! I've left what is there and hooked in the value from
the config struct (which directly replaces the env var in usage, as it also
sources from the env).

Closes #637

alamb commented on src/commands/config.rs:
```diff
@@ -1,82 +1,110 @@
+    set either with the command line flags or with the specified environment \
+    variable. If there is a file named '.env' in the current working directory, \
+    it is sourced before loading the configuration."
+)]
```
I suggest also adding the order of configuration precedence here as well (aka copy the comments from https://github.com/influxdata/influxdb_iox/pull/640/files#diff-0f0dc759b584bcbecbadd0bcd8c780a33e726fd8f6e491763febbaf92668d7c4R85 here so end users can also see it).

alamb reviewed (commented):
High level observations:
1. This PR removes the notion of a IOx config file in file that the user can specify, defaulting to `.influxdb_iox/config` and instead goes going back to using files required to be named  `.env`. I don't have a strong preference but I think given it is a user visible behavior change I would like @pauldix  to weigh in if this is OK.
2. Paul had some thoughts about a `config` command that would automatically ask the user / configure defaults in a file (e.g. ask about security tokens, etc). I am not sure how that would be achieved with this code (would it write a .env file, perhaps?) 
2. I think we could improve the usability (suggestions below) to make it clearer to a  a user what config values will be used if I run `./influxdb_iox` now.

As long as there is a well documented way to understand what the current set of config values that will be used are, I think this change is good.

Here are several suggestions / things I suggest we address:

# `.env` sourcing not working?
Something doesn't seem to be working quite right (or I don't understand how `.env` is supposed to work) as the contents of a local `.env` file don't seem to affect the config value shown:
```
alamb@ip-192-168-0-182 influxdb_iox % git status
git status
On branch dom/refactor-cli-config
Your branch is up to date with 'origin/dom/refactor-cli-config'.

nothing to commit, working tree clean
alamb@ip-192-168-0-182 influxdb_iox % cat .env 
cat .env 
INFLUXDB_IOX_BIND_ADDR=127.0.0.1:11081
alamb@ip-192-168-0-182 influxdb_iox % cargo run -- server --help
....

        --api-bind <http-bind-address>     
            The address on which IOx will serve HTTP API requests [env: INFLUXDB_IOX_BIND_ADDR=]  [default:
            127.0.0.1:8080]
...
```

I expected that the value shown for `[env: INFLUXDB_IOX_BIND_ADDR=]` would have been `127.0.0.1:11081`

# Documentation

## User docs:
Can you please also update the user documentation with the new commands? 

https://github.com/influxdata/influxdb_iox/blob/dom%2Frefactor-cli-config/README.md#L116

## output in `./influxdb_iox help`
Also, can you please update the  main help text in `src/main.rs` to  remove reference to the `config` subcommand. When I run `cargo run -- --help` the following appears (referring to `config`):

```
InfluxDB IOx server and command line tools

Examples:
    # Run the InfluxDB IOx server:
    influxdb_iox

    # Display all current config settings
    influxdb_iox config show
...
```

Perhaps we can change this to something like the following:

```
    # Display all server settings
    influxdb_iox server --help
```

alamb commented:
This is nice work @domodwyer  -- thank you. 

This is pretty cool:
<img width="213" alt="Screen Shot 2021-01-11 at 4 04 06 PM" src="https://user-images.githubusercontent.com/490673/104237831-a34aa280-5426-11eb-9453-0034c5eb12a3.png">

domodwyer commented:
Thanks @alamb - I'll tackle this between PRs shortly 👍

domodwyer commented:
I've updated the various bits - thanks for the input 👍 

Regarding the `config` command what you suggest sounds perfect - it can spit out a .env file and should all will work nicely. I've fixed the .env sourcing so it shows the values in the help text now too.

alamb commented on src/main.rs:
```diff
@@ -201,19 +206,11 @@ async fn dispatch_args(matches: ArgMatches<'_>) {
                 }
             }
         }
-        ("config", Some(sub_matches)) => {
```
I think we should remove the corresponding entry in the struct above (so the commands aren't parsed)

```
        .subcommand(
            SubCommand::with_name("config")
                .about("Configuration display and manipulation")
                .subcommand(SubCommand::with_name("show").help("show current configuration information"))
                .subcommand(SubCommand::with_name("help").help("explain detailed configuration options"))
        )
```

domodwyer replied:
Ah I didn't see that there too!

alamb reviewed (commented):
Thanks @domodwyer 

I am still having some trouble actually passing the command line arguments to the `server` subcommand. 

Specifically, I would expect to be able to start the server and pass in `--writer-id` by running `influxdb_iox server --writer-id=1`

However, when I do so I am told that it found an argument 'server' which was not valid in this context:

```
cd /Users/alamb/Software/influxdb_iox/ && cargo run -- server --writer-id=1
    Finished dev [unoptimized + debuginfo] target(s) in 0.28s
     Running `target/debug/influxdb_iox server --writer-id=1`
InfluxDB IOx server starting
error: Found argument 'server' which wasn't expected, or isn't valid in this context

USAGE:
    influxdb_iox [OPTIONS]

For more information try --help
```

Strangely, however, `influxdb_iox server --help` works:

```
influxdb_iox-server 0.1.0
Run the IOx server.

The configuration options below can be set either with the command line flags or with the specified environment
variable. If there is a file named '.env' in the current working directory, it is sourced before loading the
configuration.
    
Configuration is loaded from the following sources (highest precedence first):
        - command line arguments
        - user set environment variables
        - .env file contents
        - pre-configured default values

USAGE:
    influxdb_iox server [OPTIONS]

FLAGS:
    -h, --help       
            Prints help information

    -V, --version    
            Prints version information


OPTIONS:
        --data-dir <database-directory>    
            The location InfluxDB IOx will use to store files locally [env: INFLUXDB_IOX_DB_DIR=]  [default:
            /Users/alamb/.influxdb_iox]
        --gcp-bucket <gcp-bucket>          
            If using Google Cloud Storage for the object store, this item, as well as SERVICE_ACCOUNT must be set [env:
            INFLUXDB_IOX_GCP_BUCKET=]
        --grpc-bind <grpc-bind-address>    
            The address on which IOx will serve Storage gRPC API requests [env: INFLUXDB_IOX_GRPC_BIND_ADDR=]  [default:
            127.0.0.1:8082]
        --api-bind <http-bind-address>     
            The address on which IOx will serve HTTP API requests [env: INFLUXDB_IOX_BIND_ADDR=127.0.0.1:11081]
            [default: 127.0.0.1:8080]
        --log <rust-log>                   
            This controls the IOx server logging level, as described in https://crates.io/crates/env_logger.
            
            Levels for different modules can be specified as well. For example `debug,hyper::proto::h1=info` specifies
            debug logging for all modules except for the `hyper::proto::h1' module which will only display info level
            logging. [env: RUST_LOG=]
        --writer-id <writer-id>            
            The identifier for the server.
            
            Used for writing to object storage and as an identifier that is added to replicated writes, WAL segments and
            Chunks. Must be unique in a group of connected or semi-connected IOx servers. Must be a number that can be
            represented by a 32-bit unsigned integer. [env: INFLUXDB_IOX_ID=]
```


I also noticed that `target/debug/influxdb_iox -v` no longer works (but we can fix that as a follow on PR in my opinion): 

```
cd /Users/alamb/Software/influxdb_iox && cargo run -- -v 
    Finished dev [unoptimized + debuginfo] target(s) in 0.27s
     Running `target/debug/influxdb_iox -v`
InfluxDB IOx server starting
error: Found argument '-v' which wasn't expected, or isn't valid in this context

USAGE:
    influxdb_iox [OPTIONS]

For more information try --help
```

alamb reviewed (approved):
Its working! I vote :shipit: 

```
-*- mode: compilation; default-directory: "~/Software/influxdb_iox/" -*-
Compilation started at Tue Jan 12 09:51:32

cd /Users/alamb/Software/influxdb_iox && cargo run -- server --writer-id=1
    Finished dev [unoptimized + debuginfo] target(s) in 0.38s
     Running `target/debug/influxdb_iox server --writer-id=1`
InfluxDB IOx server starting
InfluxDB IOx server ready
```

It is really nice to be able to pass through config options via command line too auto-magically. 

The logging isn't quite working the way I expect (I can't change logging via command line argument) but I think we can fix that in a follow on PR
